braid/_braid.h

```diff
@@ -213,6 +236,7 @@ typedef struct
    braid_Real        *ta;            /**< time values                (all points) */
    braid_BaseVector  *va;            /**< restricted unknown vectors (all points, NULL on level 0) */
    braid_BaseVector  *fa;            /**< rhs vectors f              (all points, NULL on level 0) */
+   braid_Basis       *ba;            /**< time-dependent bases       (all points, NULL if not delta_correct or on level 0) */
```

Comment:
Isn't the braid_Basis already included in the braid_BaseVector ua, va, and fa?  Why do we need a separate declaration?  Is it regarding the two different bases above?  If so, we should probably specify which time-dependent basis this refers to.

Reply:
So, this is a compromise made to save on communication costs. If we wanted to avoid having to store ba, we would need some extra communication on the coarse grid. The problem is that the Delta correction needs to know the restricted fine-grid basis at the previous time-point, so that would be stored in va[i-1], which might be owned by a different processor. Currently, communication on the coarse grid only involves sending ua[-1], so having to send both ua[i-1] and the basis from va[i-1] would nearly double the communication. Maybe there is a smarter way to do this, since I am realizing that va and ba store the same data, just offset by one time-index.

Question: I know that during frestrict, on the coarse grid, the previous time-point is sent to the current processor if it is not owned, but is this value stored in va after frestrict is finished? i.e. if I tried to access va[i-1] from the step function on the coarse grid, would that work?

Reply:
It looks like `va[-1]` is updated here: https://github.com/XBraid/xbraid/blob/master/braid/restrict.c#:~:text=/*%20Initialize%20update,%7D

So, what you want may already be there.

Reply:
So this is weird, I figured I would try using the basis at va[i-1], since on one processor it should be equivalent, but for some reason it is failing to converge now. I will have to take some time to figure this out, but it looks like it should be possible to eliminate the ba variable.

Reply:
I think I've figured this out. Using va[i-1]->basis in place of ba[i] works perfectly up until fInterp, since fInterp uses va to store the error corrections, va[i-1] can be overwritten. I guess I will have to just find a way around this, but it seems doable. I'll work on this more later today.

Reply:
Sounds good...saving this communication step would be nice!

Reply:
How persistent do you need the ba[] array to be?  I thought the va[] array was potentially destroyed after interpolation (it is destroyed in an FMG cycle so I don't think the code touches it again after interpolation).  If you need a similar persistence with ba[] then maybe there is a way to restructure things so you get a valid value from va before it is overwritten (the overwriting is just to save on memory and reduce memory management overhead).

Reply:
It needs to be just as persistent as the tau correction (fa) since it is needed whenever tau correction is done. So it can be destroyed after interpolation, but it needs to persist until the end of interpolation, since the value at the current point is overwritten by the error correction, but we still need the basis vectors from the current point to step to the next point.. I'm thinking the simplest solution is to just copy the data I need to a temporary array, then use that in the step function instead of v[i-1]. That way, we are only storing a single set of basis vectors at a time, rather than one for every time-point owned by the processor.